Pull request for issues 11 and 13.

tfrank7415 opened the pull request:
Additions:
1. StoreSchemaUpgrade.cs added. This class allows users to configure store schema upgrades via the service builder in Program.cs.
![SchemaUpgradeExample](https://user-images.githubusercontent.com/44216318/210026517-a0b560f5-1e84-47fc-bec2-1b0b611eb098.PNG)
2. IndexedDbUpgradeFunctions.cs added. This class defines what different upgrade actions the user can perform.
![IndexedDbUpgradeFunctions](https://user-images.githubusercontent.com/44216318/210026619-b3b27313-126c-4725-9562-2e56f366174a.PNG) 
3. Slight change to blazorDB.createDB() to check if the user wants to perform a schema upgrade. Code is commented, so you can review.
4. 10 functions added to blazorDB.js to perform the schema upgrades. 
5. Updated example project.
6. 5 functions added to IndexDbManager.cs and blazorDB.js to address issue 13. New functions allow the user retrieve the first item in a store, the last item in a store, the min index for a store, the max index for a store, and the count of the records in a store.

Commit:
Check-in for Issue 11. 10 functions added to blazorDb.js that allow users to make schema upgrades.

Commit:
Cleaning up excess code in Program.cs.

tfrank7415 commented:
Working on issue 13 next. Let me know if you have any coding preferences, so I can implement them on my next pull request.

Commit:
5 new functions added to address issue #13.

Commit:
Added comments.

nwestfall commented on src/BlazorDB/StoreSchemaUpgrade.cs:
```diff
@@ -0,0 +1,23 @@
+    /// </summary>
+    public class StoreSchemaUpgrade
+    {
+        // Store to upgrade
+        public string Name { get; set; }
+        // Upgrade action to perform
+        public string UpgradeAction { get; set; }
+        // Object to contain parameters needed to complete the action
+        public object[] UpgradeActionParameterList { get; set; }
+        // List of column to perform action AND/OR delete
+        public List<string> ColumnsToPerformActionOn { get; set; }
+        // Column to receive data after action is performed
+        public List<string> ColumnsToReceiveDataFromAction { get; set; }
```
You can make these C# comments

```xml
<summary>
Upgrade action to perform
</summary>
```

tfrank7415 replied:
Implemented.

nwestfall commented on src/BlazorDB/DbStore.cs:
```diff
@@ -7,5 +7,9 @@ public class DbStore
         public int Version { get; set; }
         public List<StoreSchema> StoreSchemas { get; set; }
+
+        // New Code for Fork: BlazorDB-issue-11
+        public List<StoreSchemaUpgrade> StoreSchemaUpgrades { get; set; }
+        // End New Code for Fork: BlazorDB-issue-11
```
Remove comments

tfrank7415 replied:
Implemented.

nwestfall commented on src/BlazorDB/IndexedDbUpgradeFunctions.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    public struct IndexedDbUpgradeFunctions
+    {
+        // Splits the data in one column into two new columns
+        // Must include two column names in the ColumnsToReceiveDataFromAction property
+        public const string SPLIT_COLUMN = "split";
+        // Multiplies the value of one column, puts the new value in the same column
+        // ColumnsToPerformActionOn AND ColumnsToReceiveDataFromAction should be the same
+        public const string MULTIPLY_COLUMN = "multiply";
+        // Multiplies the value of one column, puts new value in new column, deletes old column.
+        // ColumnsToPerformActionOn is the column you want to multiply AND the column that will be deleted
+        // ColumnsToReceiveDataFromAction is the new column that will receive the new value
+        public const string MULTIPLY_COLUMN_DELETE = "multiply-delete";
+        // Divides the value of one column
+        // ColumnsToPerformActionOn should contain the name of the column you want to divide
+        public const string DIVIDE_COLUMN = "divide";
+        // Divides the value of one column, puts new value in new column, deletes old column.
+        // ColumnsToPerformActionOn should contain the name of the column you want to divide AND delete
+        // ColumnsToReceiveDataFromAction should contain the name of column you want to receive the new value
+        public const string DIVIDE_COLUMN_DELETE = "divide-delete";
+        // Adds given value to column
+        // ColumnsToPerformActionOn should be the name of the column that contains the value you want to add to
+        // ColumnsToReceiveDataFromAction should be the same name as ColumnsToPerformActionOn
+        public const string ADD_COLUMN = "add";
+        // Adds given value to one column, puts new value in new column, deletes old column.
+        // ColumnsToPerformActionOn should be the name of the column that contains the value you want to add to AND delete
+        // ColumnsToReceiveDataFromAction should the column you want to insert the new value into
+        public const string ADD_COLUMN_DELETE = "add-delete";
+        // Subtracts given value from one column
+        // ColumnsToPerformActionOn should be the column you want to subtract from
+        // ColumnsToReceiveDataFromAction should be the same as ColumnsToPerformActionOn
+        public const string SUBTRACT_COLUMN = "subtract";
+        // Subtracts given value from one column, puts new value in new column, deletes old column.
+        // ColumnsToPerformActionOn should be the column you want to subtract from AND delete
+        // ColumnsToReceiveDataFromAction should be the column you want the new value to be inserted into
+        public const string SUBTRACT_COLUMN_DELETE = "subtract-delete";
+        // Deletes column
+        // ColumnsToPerformActionOn is the column you want to delete
+        // ColumnsToReceiveDataFromAction should not have a value
+        public const string DELETE_COLUMN = "delete-column";
```
Make these C# comments

```xml
<summary>
Deletes column

ColumnsToPerformActionOn is the column you want to delete
ColumnsToReceiveDataFromAction should not have a value
</summary>
```

tfrank7415 replied:
Implemented.

nwestfall commented on src/BlazorDB/wwwroot/blazorDB.js:
```diff
@@ -149,6 +161,74 @@ window.blazorDB = {
+            window.blazorDB.getTable(item.dbName, item.storeName).then(table => {
+                table.toCollection().count().then((result) => {
+                    dotnetReference.invokeMethodAsync('BlazorDBCallback', transaction, false, 'getMinIndex succeeded');
+                    resolve(result == 0 ? "Table empty" : 0);
```
What about returning a `-1` instead of `Table Empty`?

The concern is just getting it the same type return for `if/else`

tfrank7415 replied:
Implemented.

nwestfall commented on src/BlazorDB/wwwroot/blazorDB.js:
```diff
@@ -149,6 +161,74 @@ window.blazorDB = {
+            window.blazorDB.getTable(item.dbName, item.storeName).then(table => {
+                table.toCollection().count().then((result) => {
+                    dotnetReference.invokeMethodAsync('BlazorDBCallback', transaction, false, 'getMaxIndex succeeded');
+                    resolve(result > 0 ? (result - 1) : "Table empty");
```
What about returning a `-1` instead of `Table Empty`?

The concern is just getting it the same type return for `if/else`

tfrank7415 replied:
Implemented.

nwestfall commented on src/BlazorDB/wwwroot/blazorDB.js:
Just noting this assumption.

I'm okay for a round 1 draft too, it's just one update per version and we just error if `storeSchemeUpgrades.length > 1` to let the user know.

nwestfall commented on src/BlazorDB/wwwroot/blazorDB.js:
I'd error with something like `Upgrade not implemented` and callback to C# if possible. The C# callback is only if it's easy here, not required.

tfrank7415 replied:
Warning implemented. C# callback not implemented. I will research this and try to implement it on the next iteration.

nwestfall commented on src/BlazorDB/wwwroot/blazorDB.js:
This is a great add, only thing I'd look at doing is handle some possible user errors.  Some that come to mind

  - Column in row doesn't exist
  - If the column for the row returns a `null` value how does `split` or other actions perform?
  
  Just looking to see how we support these cases

tfrank7415 replied:
Working this now. I'll provide more feedback later tonight or tomorrow.

tfrank7415 replied:
Currently, all errors are caught in the catch at line 63 in blazorDB.js. So if I try to split a row that does not exist, the following error is logged to the console. 
![TypeError](https://user-images.githubusercontent.com/44216318/210296135-ca8a5a4e-9e71-49a5-9f0a-d5aa179e8263.PNG)

We need to make these errors more specific and useful. I think the best way to do that is to create an error handling function and call it inside of upgradeDbSchema(). The error handling function will look at the schemaUpdate.upgradeAction property and perform different checks to ensure all properties contain useable data. If a property does not contain useful data, I will throw an error message relative to that property. For instance, if the schemaUpdate.upgradeAction == 'split' and the columnsToPerformActionOn property is null or does not exist in the store I will throw an error that states, "columnsToPerformActionOn is null" or "columnsToPerformActionOn does not exist in store". 

This thrown error will be caught at line 63 in blazorDB.js and logged to the console. The upgrade error will not effect the db instance since the version.upgrade() creates a database transaction. Dexie.js will simply use the last version of the database. So whatever the current version is. 

What do you think? I think the custom error handling function will be the only way to return useful error messages.

tfrank7415 replied:
Change of plans after testing. I decided to add verifyStoreSchemaUpgrades() to the if logic that determines if upgradeDbSchema() is called. See below. 
![IfLogic](https://user-images.githubusercontent.com/44216318/210480360-191b693e-5445-4e9b-81ac-2e8722064202.PNG)

This ensures we will not attempt to perform a database upgrade via upgradeDbSchema() with incorrect parameters. I like this change because it ensures that the opening of a database will not be affected by any error that occurs in upgradeDbSchema().

verifyStoreSchemaUpgrade() will return false if the give storeSchemaUpgrade parameters do not match the given schemaUpdate.upgradeAction. See below for an example. 
![VerifyStoreSchemaUpgrade](https://user-images.githubusercontent.com/44216318/210480842-994dfca4-93ff-4d71-b47e-80c155756214.PNG)

I will finish verifyStoreSchemaUpgrades() tomorrow then commit my code. Let me know if you have any feedback.

nwestfall replied:
I like this approach.  Nice change.

tfrank7415 replied:
Sweet. Let me know if you see anything else I need to change on my end.

nwestfall commented on BlazorDB.sln:
```diff
@@ -1,11 +1,11 @@
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "BlazorDB", "src\BlazorDB\BlazorDB.csproj", "{D0180843-D92A-44B7-9893-F2450415F208}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "BlazorDB.Example", "example\BlazorDB.Example\BlazorDB.Example.csproj", "{155CD782-6CEF-4115-9A04-1864A2CDFC90}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "BlazorDB.Example", "BlazorDB.Example\BlazorDB.Example.csproj", "{A47C9FCB-70BC-4B16-B5BB-68C8F6B1012E}"
```
Just curious why we are moving this out of the `example` folder?

tfrank7415 replied:
This was unintentional. Will be fixed on next commit.

nwestfall reviewed (changes requested):
Nice!  Overall, this is really solid.  Let me know what you think of the comments and we can go from there.

Commit:
Requested changes for PR 19.

Commit:
Added verifyStoreSchemaUpgrades() to blazorDB.js.

tfrank7415 commented:
Committed latest changes. Main change is that I added verifyStoreSchemaUpgrades() to blazorDB.js. I also added it to the if logic that determines if upgradeDbSchema() is called. See below. 
![IfLogic](https://user-images.githubusercontent.com/44216318/210706455-f67b5d42-d495-4f88-9d0c-b10229741a52.PNG)

verifyStoreSchemaUpgrades() will return false and console.error() if the following conditions occurs:
1. columnsToPerformActionOn, columnsToReceiveDataFromAction, and upgradeActionParameterList are not the expected length. This means they do not contain the correct number of column names. 
2. columnsToPerformActionOn, columnsToReceiveDataFromAction, and upgradeActionParameterList are null or empty.

The above checks ensure that the database will always be opened even if the user creates an incorrect StoreSchemaUpgrade object in Program.cs. 

Caveat: I could not verify that the column names passed by the property columnsToPerformActionOn were existing column names in a store. This is because I do not have access to Dexie.db.tables until after db.open() is called. Also, only indexed columns are stored in the store schema. So even if I had the ability to use db.tables, I would only be able to check the indexed columns to see if they existed.
